lib/PhasarLLVM/ControlFlow/LLVMBasedICFG.cpp

```diff
@@ -751,6 +757,198 @@ nlohmann::json LLVMBasedICFG::getAsJson() const {
 
 void LLVMBasedICFG::printAsJson(std::ostream &OS) const { OS << getAsJson(); }
 
+nlohmann::json LLVMBasedICFG::exportICFGAsJson() const {
+  nlohmann::json J;
+
+  llvm::DenseSet<const llvm::Instruction *> handledCallSites;
+
+  for (const auto *F : getAllFunctions()) {
+    for (auto &[From, To] : getAllControlFlowEdges(F)) {
+      if (llvm::isa<llvm::UnreachableInst>(From)) {
+        continue;
+      }
+
+      if (const auto *Call = llvm::dyn_cast<llvm::CallBase>(From)) {
+        auto [unused, inserted] = handledCallSites.insert(Call);
+
+        for (const auto *Callee : getCalleesOfCallAt(Call)) {
+          if (Callee->isDeclaration()) {
+            continue;
+          }
+          if (inserted) {
+            J.push_back({{"from", llvmIRToString(From)},
+                         {"to", llvmIRToString(&Callee->front().front())}});
+          }
+
+          for (const auto *ExitInst : getAllExitStatements(Callee)) {
+            J.push_back({{"from", llvmIRToString(ExitInst)},
+                         {"to", llvmIRToString(To)}});
+          }
+        }
+
+      } else {
+        J.push_back(
+            {{"from", llvmIRToString(From)}, {"to", llvmIRToString(To)}});
+      }
+    }
+  }
+
+  return J;
+}
+
+struct SourceCodeInfoWithIR : public SourceCodeInfo {
+  std::string IR;
+};
+
+void from_json(const nlohmann::json &J, SourceCodeInfoWithIR &Info) {
+  from_json(J, static_cast<SourceCodeInfo &>(Info));
+  J.at("IR").get_to(Info.IR);
+}
+void to_json(nlohmann::json &J, const SourceCodeInfoWithIR &Info) {
+  to_json(J, static_cast<const SourceCodeInfo &>(Info));
+  J["IR"] = Info.IR;
+}
```

Comment:
We probably should put these functions in an anonymous namespace, right?

Reply:
nlohmann/json requires these functions to be in the same namespace in order to automatically call them. Otherwise, it will pick the ones from the base class (`psr::SourceCodeInfo`)